File: job-agent-os/src/job_agent_os/schemas/session.py
"""Session schemas."""

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, Field


class SessionCreate(BaseModel):
    """Session creation request."""

    intent: str
    mode: str = Field(default="full", pattern="^(full|search_only|match_only|resume_only|interview_only)$")
    resume_id: UUID | None = None
    options: dict = Field(default_factory=dict)


class SessionProgress(BaseModel):
    """Session progress info."""

    completed_steps: list[str] = Field(default_factory=list)
    current_step: str | None = None
    pending_steps: list[str] = Field(default_factory=list)
    # Updated by HarnessRuntime when a graph node starts/finishes.  Keeping
    # this separate from ``current_step`` lets clients state plainly which
    # Agent is working without having to infer it from a pipeline position.
    active_agent: str | None = None
    active_agent_status: str | None = None
    # Fine-grained, user-safe activity published by long-running search work.
    # These are intentionally aggregate values; source URLs and raw crawler
    # output never belong in an SSE/session progress payload.
    activity_message: str | None = None
    items_found: int = 0
    items_saved: int = 0
    last_activity_at: str | None = None


class PendingApprovalInfo(BaseModel):
    """Pending approval info."""

    id: UUID
    type: str
    title: str


class SessionResponse(BaseModel):
    """Session response."""

    session_id: UUID
    status: str
    intent: str | None = None
    current_phase: str | None = None
    progress: SessionProgress | None = None
    pending_approval: PendingApprovalInfo | None = None
    results_summary: dict = Field(default_factory=dict)
    token_usage: dict = Field(default_factory=dict)
    started_at: datetime | None = None
    updated_at: datetime | None = None


class SessionMessage(BaseModel):
    """Session message request."""

    content: str
    message_type: str = Field(default="text", pattern="^(text|clarification|feedback)$")


class SessionMessageResponse(BaseModel):
    """Session message response."""

    message_id: UUID
    response: str
    session_status: str

File: job-agent-os/frontend/src/types/session.ts
/** Session types (matches backend schemas/session.py) */

export type SessionMode =
  | "full"
  | "search_only"
  | "match_only"
  | "resume_only"
  | "interview_only";

export interface SessionCreateRequest {
  intent: string;
  mode?: SessionMode;
  resume_id?: string | null;
  options?: Record<string, unknown>;
}

export interface SessionProgress {
  completed_steps: string[];
  current_step: string | null;
  pending_steps: string[];
  /** The graph node currently executing, published by the backend runtime. */
  active_agent?: string | null;
  active_agent_status?: "running" | "completed" | "failed" | "waiting" | null;
  /** Optional crawl-level activity details, when the active Agent exposes them. */
  activity_message?: string | null;
  items_found?: number | null;
  items_saved?: number | null;
  last_activity_at?: string | null;
}

export interface PendingApprovalInfo {
  id: string;
  type: string;
  title: string;
}

export type SessionStatus =
  | "created"
  | "running"
  | "waiting_approval"
  | "waiting_input"
  | "completed"
  | "failed"
  | "cancelled";

export interface SessionResponse {
  session_id: string;
  status: SessionStatus;
  intent: string | null;
  current_phase: string | null;
  progress: SessionProgress | null;
  pending_approval: PendingApprovalInfo | null;
  results_summary: Record<string, unknown>;
  token_usage: Record<string, number>;
  started_at: string | null;
  updated_at: string | null;
}

export interface SessionStreamEvent {
  phase?: string | null;
  status?: SessionStatus;
  progress?: SessionProgress | null;
  done?: boolean;
  reason?: string;
  clarification_question?: string | null;
}

export type MessageType = "text" | "clarification" | "feedback";

export interface SessionMessageRequest {
  content: string;
  message_type?: MessageType;
}

export interface SessionMessageResponse {
  message_id: string;
  response: string;
  session_status: SessionStatus;
}

/** Timeline event (from GET /v1/sessions/{id}/timeline) */
export interface TimelineEvent {
  event: string;
  timestamp: string | null;
  data: Record<string, unknown>;
}

/** Session execution log entry (from GET /v1/sessions/{id}/logs) */
export interface SessionLogEntry {
  id: string;
  agent_name: string;
  node_name: string;
  step_order: number;
  status: string;
  duration_ms: number | null;
  total_tokens: number | null;
  error_message: string | null;
  started_at: string | null;
  finished_at: string | null;
}

/** Resume diff entry (from resume optimization approval payload) */
export interface ResumeDiffEntry {
  section?: string;
  before: string;
  after: string;
  reason: string;
}

/** Interview question (matches backend schema and interview API type) */
export interface InterviewQuestion {
  id: number;
  type: string;
  category: string;
  difficulty: string;
  question: string;
  reference_answer: string;
  scoring_criteria: string[];
  follow_up: string | null;
}
